public/app/features/explore/utils/decorators.ts

```diff
@@ -1,214 +1,187 @@
-import { MonoTypeOperatorFunction, of, OperatorFunction } from 'rxjs';
-import { map, mergeMap } from 'rxjs/operators';
+import { Observable, of } from 'rxjs';
+import { map } from 'rxjs/operators';
 import {
+  AbsoluteTimeRange,
   DataFrame,
-  DataSourceApi,
   FieldType,
   getDisplayProcessor,
   PanelData,
-  PreferredVisualisationType,
   sortLogsResult,
   standardTransformers,
 } from '@grafana/data';
 import { config } from '@grafana/runtime';
+import { groupBy } from 'lodash';
 
-import { ExploreItemState, ExplorePanelData } from '../../../types';
+import { ExplorePanelData } from '../../../types';
 import { getGraphSeriesModel } from '../../../plugins/panel/graph2/getGraphSeriesModel';
 import { dataFrameToLogsModel } from '../../../core/logs_model';
 import { refreshIntervalToSortOrder } from '../../../core/utils/explore';
 
-export const decorateWithGraphLogsTraceAndTable = (
-  datasourceInstance?: DataSourceApi | null
-): OperatorFunction<PanelData, ExplorePanelData> => inputStream =>
-  inputStream.pipe(
-    map(data => {
-      if (data.error) {
-        return {
-          ...data,
-          graphFrames: [],
-          tableFrames: [],
-          logsFrames: [],
-          traceFrames: [],
-          graphResult: null,
-          tableResult: null,
-          logsResult: null,
-        };
-      }
-
-      const graphFrames: DataFrame[] = [];
-      const tableFrames: DataFrame[] = [];
-      const logsFrames: DataFrame[] = [];
-      const traceFrames: DataFrame[] = [];
+/**
+ * When processing response first we try to determine what kind of dataframes we got as one query can return multiple
+ * dataFrames with different type of data. This is later used for type specific processing. As we use this in
+ * Observable pipeline, it decorates the existing panelData to pass the results to later processing stages.
+ */
+export const decorateWithGraphLogsTraceAndTable = (data: PanelData): ExplorePanelData => {
```

Comment:
Functions here were changed to just by a mapper functions. That should make it easier to test and reuse if needed with only tradeoff that we have to use them with `map()` when using with observables.

Reply:
This simplifies the tests because it removes RxJs and observableTester right?

Reply:
Yes. The table mapper still returns observable due to the transformer so that needs to be handled, but others don't do anything async so could be used as simple mappers. It's not a big change and I mostly done it as I was exploring how the code works right now. Also the table transform could probably use toArray and then toPromise in case it emits multiple times but I assume it's not necessary at the moment.